src/applications/microservices/payforadoption-go/sql_span_processor.go

```diff
@@ -0,0 +1,162 @@
+
+	// Check for known SQL operations
+	for prefix, operation := range sqlOperations {
+		if strings.HasPrefix(sqlUpper, prefix) {
+			return operation
+		}
+	}
+
+	return unknownRemoteOperation
+}
+
+// Shutdown is called when the processor is shut down
+func (p *SQLSpanProcessor) Shutdown(ctx context.Context) error {
+	return nil
+}
+
+// ForceFlush is called to force flush any buffered spans
+func (p *SQLSpanProcessor) ForceFlush(ctx context.Context) error {
+	return nil
+}
+
+// isDatabaseSpan checks if the span is a database span that needs Aurora correlation
+func (p *SQLSpanProcessor) isDatabaseSpan(spanName string, s trace.ReadWriteSpan) bool {
+	// Check if span name indicates a database operation
+	if spanName == dbSystemPostgres {
+		return true
+	}
+
+	// Also check for db.system attribute to identify database spans
+	for _, attr := range s.Attributes() {
+		if attr.Key == "db.system" && attr.Value.AsString() == dbSystemPostgres {
+			return true
+		}
+	}
+
+	return false
+}
+
+// buildAuroraCorrelationAttributes builds the Aurora correlation attributes for a span
+func (p *SQLSpanProcessor) buildAuroraCorrelationAttributes(sqlOperation string) []attribute.KeyValue {
+	return []attribute.KeyValue{
+		// Ensure the service name is preserved
+		attribute.String("aws.remote.service", dbSystemPostgres),
+		// Add Aurora correlation attributes
+		attribute.String("aws.remote.resource.identifier", p.resourceIdentifier),
+		attribute.String("aws.remote.resource.type", "DB::Connection"),
+		attribute.String("aws.remote.operation", sqlOperation),
+		attribute.String("remote.db.user", p.dbUser),
+		attribute.String("remote.resource.cfn.primary.identifier", p.resourceIdentifier),
+		// Add database connection string for correlation (sanitized)
+		attribute.String("db.connection_string", dbConnectionString),
+	}
+}
+
+// extractSQLOperationFromSpan extracts the SQL operation from span attributes
+func (p *SQLSpanProcessor) extractSQLOperationFromSpan(s trace.ReadWriteSpan) string {
+	// Look for db.statement attribute which contains the SQL query
+	for _, attr := range s.Attributes() {
+		if attr.Key == "db.statement" {
+			sqlStatement := attr.Value.AsString()
+			return p.extractSQLOperation(sqlStatement)
+		}
+	}
+
+	// If no SQL statement found, return unknown operation
+	return unknownRemoteOperation
+}
+
+// createSQLSpanProcessor creates and configures the SQL span processor
+func createSQLSpanProcessor(ctx context.Context, cfg payforadoption.Config) (*SQLSpanProcessor, error) {
+	// Get database configuration for correlation attributes
+	dbService := payforadoption.NewDatabaseConfigService(cfg)
+	dbConfig, err := dbService.GetDatabaseConfig(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get database config for SQL span processor: %w", err)
+	}
+
+	// Build resource identifier for Aurora correlation
+	resourceIdentifier := fmt.Sprintf("postgres|%s|%d", dbConfig.Host, dbConfig.Port)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Potential log injection detected. Ensure all untrusted input is properly sanitized before logging. Use parameterized logging or validate input against an allow list to prevent log injection vulnerabilities. Consider using a dedicated logging library&#39;s built-in sanitization features when available. Learn more - https://cwe.mitre.org/data/definitions/117.html